### PR TITLE
Working on DeepQAgent

### PR DESCRIPTION
**Pull request by Neo-David**
Creating a nn model that works
for the DeepQAgent. Also adjusting
the gym training to work with this.
added scripts to contiue training.

### CODE REVIEW

File: AI/training_script.py [hunk 1]
```diff
@@ -4,18 +4,27 @@
...
-gym.train()
+gym.train(num_episodes=training_iterations)
 gym.eval_exp_table()
 
 # d) Run test
-gym.test()
+gym.test(10000)
 gym.eval()
+
+gym.save_model("dq_agent.h5")
```
Comment (dstaahl) [inline AI/training_script.py]: Let the player save the model here

File: AI/pyneogame/Agent/DeepQAgent.py [hunk 2]
```diff
@@ -26,28 +28,30 @@ def __init__(self, state_size, actions,
                  model=None,
                  epsilon=0.9,
                  decay_rate=1e-5,
-                 update_interval=200,
-                 memory_size=10000,
+                 update_interval=25000,
```
Comment (johan-sjogren) [inline AI/pyneogame/Agent/DeepQAgent.py]: Det borde nog ha varit en egen klass som ärver från DQ agent.

File: AI/training_script.py [hunk 3]
```diff
@@ -4,18 +4,27 @@
 from pyneogame.gym import Gym
 from pyneogame.Agent.QTableAgent import QTableAgent
 from pyneogame.Agent.GreedyAgent import GreedyAgent
+from pyneogame.Agent.DeepQAgent import DeepQAgent
+from pyneogame.Engine import Game
 
 # a) Setup the agents
-player = QTableAgent(unexplored=1)
+
+training_iterations = 100000
```
Comment (dstaahl) [inline AI/training_script.py]: Uppercase for global variables

File: AI/training_script.py [hunk 4]
```diff
@@ -4,18 +4,27 @@
...
+update_interval = training_iterations / 5
```
Comment (dstaahl) [inline AI/training_script.py]: Set to 5 and use modulo operation in gym class

File: AI/pyneogame/gym.py [hunk 5]
```diff
@@ -189,3 +198,10 @@ def eval(self):
         print("\tWins   {0:.2f}%".format(100.0 * ratio_player_win / ratio_optimal_win))
         print("\tLosses {0:.2f}%".format(100.0 * ratio_opponent_win / ratio_optimal_loose))
         print("\tScore {0:.2f}%".format(100.0 * relative_result))
+
+        if self.last_test is not None:
+            print("Diff from last test score is {0:.2f}%".format(100.0 * (relative_result - self.last_test)))
+        self.last_test = relative_result
+
+    def save_model(self, file_name):
```
Comment (dstaahl) [inline AI/pyneogame/gym.py]: Let the player save its model

File: AI/pyneogame/gym.py [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
 class Gym:
 
-    def __init__(self, player, opponent):
+    def __init__(self, player, opponent, name="autosave_model.h5"):
```
Comment (dstaahl) [inline AI/pyneogame/gym.py]: Good!

File: AI/pyneogame/gym.py [hunk 7]
```diff
@@ -97,6 +98,14 @@ def train(self, num_episodes=10000):
             self.player.learn(player_state,
                          player_action,
                          reward)
+            self.player.learn(opponent_state,
+                              opponent_action,
+                              -reward)
+            if self.player.episode_counter == 0:
```
Comment (dstaahl) [inline AI/pyneogame/gym.py]: ?

File: AI/pyneogame/gym.py (hunk 6 above)
Comment (dstaahl) [inline AI/pyneogame/gym.py]: Change class attribute 'name' to  'filename' or similar and add checkpoints with suffix '_chkpt<number>'

**Pull request closed without merging**